### ISSUE TITLE
Track development of Chapel module for Apache Arrow

### ISSUE THREAD

**Issue by mhmerrill** [Bears-R-Us/arkouda#833]
This issue is a place holder to track development of Chapel module for Apache Arrow.

**Comment by ShreyasKhandekar** [comment]
I am currently working on adding support for Apache Arrow
and Parquet in Chapel.
In order to add functionality for these two libraries in
Chapel, I decided to first familiarize myself with them
and their salient features.
I have written Python programs that deal with Arrow data
structures like record batches and tables across
different processes and formats. I have also tried
reading and Parquet files and converting them to Arrow
data structures and vice-versa.
Then I recreated the same in C.

Starting with python gave me the ability to pick up both
libraries quickly and transliterating to C was an
important step to getting towards doing the same in
Chapel.
From here I will take Chapel's interoperability with C in
order to thinly wrap the C interface of Apache Arrow
(which also includes support for dealing with Apache
Parquet files and deserializing them into Arrow data
structures) in Chapel.

Doing so will leverage the existing C implementation of
Apache Arrow and Parquet combined with Chapel's
interoperability with C.

The next step would be to make the implementation more
abstract as the C implementation is verbose and exposes a
lot of how Arrow actually represents the data in memory.
This would mean creating a higher level Chapeltastic
interface that makes better use of Chapel Types and
features by wrapping and calling down to the C-level
interface.

**Comment by hokiegeek2** [comment]
@ShreyasKhandekar AWESOME!

**Comment by bradcray** [comment]
I'm heading out for vacation and cleaning my desk, and am noticing that @ShreyasKhandekar didn't post an update here that I think he'd intended to.  I've encouraged him to post it when he's back in the office Monday, but the quick summary is that he's gotten his sample C programs transliterated over to Chapel using a C-level interface.  The next step is to raise the level of abstraction in order to make it more succinct and Chapeltastic (e.g., hopefully reducing 200-line sequences down to 5-lines, more similar to Python).

**Comment by ShreyasKhandekar** [comment]
Hey everyone, thanks for waiting for this update!
We made some more progress with this and have successfully made some calls to Arrow from Chapel.
As I laid out in my last post, I transcribed a subset of the Arrow library into Chapel and tried some basic 
operations like creating an Arrow record batch or reading Parquet files using those functions which in turn call the C implementation of Arrow.
Though we still need to wrap the rest of the C implementation in Chapel, this is a great milestone in using Arrow in Chapel.
But since this interface is still pretty verbose and hard to write and understand, we have begun thinking about what the design of the more abstract/higher-level Chapeltastic interface would look like. Once we have a better idea about that, I will recreate the same calls I made using the verbose interface and contrast them with how the high-level interface might do the same, but in an easier fashion.
